### ISSUE TITLE
Skill: (1) Allow users to sign up for new accounts in Rails

### ISSUE THREAD

**Issue by PStanos** [PStanos/tc359_rails_final#7]

**Comment by PStanos** [comment]
New users can be created if you are not currently logged in. The "New User" link is part of the template for the entire app. @chrisvfritz

**Comment by chrisvfritz** [comment]
There are a lot of things you've done here that demonstrate a growing understanding of Rails. Well done! :smiley: There's still quite a bit of room for improvement though, so I've performed a refactor described in the diff below that I'd like you to try to implement. I've often included comments to explain why I made changes, but please don't hesitate to let me know if something doesn't make sense to you or if you'd like more information. Hopefully it will help even better understand some of Rails' magic!

``` diff
diff --git a/app/controllers/sessions_controller.rb b/app/controllers/sessions_controller.rb
index fa3f4fa..46b86be 100644
--- a/app/controllers/sessions_controller.rb
+++ b/app/controllers/sessions_controller.rb
@@ -1,14 +1,15 @@
 class SessionsController < ApplicationController
-  include SessionsHelper

   def new
   end

   def create
     user = User.find_by(email: params[:email])
-    if login(user)
+    if user && user.authenticate(params[:password])
+      session[:user_id] = user.id
       redirect_to root_url, notice: 'Logged in!'
     else
+      flash.alert = 'Invalid email/password combination. Please try again.'
       render :new
     end
   end
@@ -17,4 +18,5 @@ class SessionsController < ApplicationController
     session[:user_id] = nil
     redirect_to root_url, notice: 'Successfully signed out!'
   end
+
 end

diff --git a/app/controllers/users_controller.rb b/app/controllers/users_controller.rb
index 7098717..31ebc98 100644
--- a/app/controllers/users_controller.rb
+++ b/app/controllers/users_controller.rb
@@ -1,29 +1,29 @@
 class UsersController < ApplicationController
-  include SessionsHelper

   def new
-    render :new_user
+    @user = User.new
+    # We don't need to specify a render when the view is also named "new"
   end

   def create
-    if(params[:password] != params[:repeat_password])
-      flash.alert = 'Passwords do not match. Please try again.'
-      render :new_user
-      return
-    end
-
-    @user = User.new(email: params[:email], password: params[:password])
+    @user = User.new(user_params)

-   if !@user.save
-      render :new_user
-      return
-    end
-
-    if login(@user)
+    if @user.save
+      session[:user_id] = @user.id
       redirect_to root_url, notice: 'Account created!'
     else
-      flash.alert = 'User creation failed. Please try again.'
-      render :new_user
+      flash.alert = 'Registration failed. Please try again.'
+      render :new
     end
   end
+
+# The helpers modules are typically meant for the views. In controllers, where we can organize things in a class, it's more typical to have a private section at the end of the class, listing non-action methods we want to use after it.
+
+private
+
+  # Rails has a feature calld "strong parameters" which handles attribute whitelisting like this:
+  def user_params
+    params.require(:user).permit(:email, :password, :password_confirmation)
+  end
+
 end
\ No newline at end of file
diff --git a/app/helpers/sessions_helper.rb b/app/helpers/sessions_helper.rb
index 26e0816..309f8b2 100644
--- a/app/helpers/sessions_helper.rb
+++ b/app/helpers/sessions_helper.rb
@@ -1,11 +1,2 @@
 module SessionsHelper
-  def login(user)
-    if user && user.authenticate(params[:password])
-      session[:user_id] = user.id
-      return true
-    else
-      flash.alert = 'Invalid email/password combination. Please try again.'
-      return false
-    end
-  end
 end
diff --git a/app/views/layouts/application.slim b/app/views/layouts/application.slim
index 0a85745..a78363a 100644
--- a/app/views/layouts/application.slim
+++ b/app/views/layouts/application.slim
@@ -19,7 +19,7 @@ html
           = link_to "Sign Out", sign_out_path, method: :delete
       - else
         li
-          = link_to "New User", new_user_path
+          = link_to "Register", new_user_path
         li
           = link_to "Sign In", sign_in_path
     == yield
diff --git a/app/views/sessions/new.slim b/app/views/sessions/new.slim
index d25a12e..b8bd2b7 100644
--- a/app/views/sessions/new.slim
+++ b/app/views/sessions/new.slim
@@ -1,13 +1,13 @@
-h1
-  | Log In
+h1 Log In
+
 = form_tag sessions_path do
-  div class="field"
+  .field
     = label_tag :email
     br
     = text_field_tag :email
-  div class="field"
+  .field
     = label_tag :password
     br
     = password_field_tag :password
-  div class="actions"
-    = submit_tag "Log In"
\ No newline at end of file
+  .actions
+    = submit_tag 'Log In'
\ No newline at end of file
diff --git a/app/views/static/home.slim b/app/views/static/home.slim
index e100b25..e8d44a8 100644
--- a/app/views/static/home.slim
+++ b/app/views/static/home.slim
@@ -1,3 +1,2 @@
-h1
-  'Welcome to the Animal Shelter Website
+h1 Welcome to the Animal Shelter Website
 = link_to 'Cats at the Animal Shelter', cats_path
\ No newline at end of file
diff --git a/app/views/users/_form.slim b/app/views/users/_form.slim
index 3cdd06e..1475a24 100644
--- a/app/views/users/_form.slim
+++ b/app/views/users/_form.slim
@@ -1,25 +1,31 @@
-= form_tag("/new_user/create", method: "get")
+/ Use form_for instead of form_tag for objects, as it allows you to use more concise helpers
+
+= form_for @user do |f|
+
   - if @user and @user.errors.any?
-      div id="error_explanation"
-        h2
-          = pluralize(@user.errors.count, "error")
-            | prohibited this user from being created:
+    / You can shorten div definitions in Slim by just listing the ids and/or classes on that div, similarly to how you would in target the same element in CSS
+    #error_explanation
+      / You can use #{} to mix Ruby into plain text in Slim
+      h2 #{ pluralize @user.errors.count, 'error' } prohibited this user from being created:
+      ul
+        - @user.errors.full_messages.each do |message|
+          li= message
+
+  .field
+    = f.label :email
+    = f.text_field :email
+
+  .field
+    = f.label :password
+    = f.password_field :password
+
+  / Calling this field password_confirmation takes an advantage of a feature given to us by
+  / has_secure_password, which we're using in our User model. It will automatically check
+  / whether the confirmation is the same as the password.
+  .field
+    = f.label :password_confirmation
+    = f.password_field :password_confirmation

-        ul
-          - @user.errors.full_messages.each do |message|
-            li
-              = message
-  div class="field"
-    = label_tag :email
-    br
-    = text_field_tag :email
-  div class="field"
-    = label_tag :password
-    br
-    = password_field_tag :password
-  div class="field"
-    = label_tag :repeat_password
-    br
-    = password_field_tag :repeat_password
-  div class="actions"
-    = submit_tag("Create")
\ No newline at end of file
+  .actions
+    / I recommind using language like 'Register' or 'Sign up' instead of 'Create user'
+    = f.submit 'Register'
\ No newline at end of file
diff --git a/app/views/users/new_user.slim b/app/views/users/new_user.slim
deleted file mode 100644
-    br
-    = text_field_tag :email
-  div class="field"
-    = label_tag :password
-    br
-    = password_field_tag :password
-  div class="field"
-    = label_tag :repeat_password
-    br
-    = password_field_tag :repeat_password
-  div class="actions"
-    = submit_tag("Create")
\ No newline at end of file
+  .actions
+    / I recommind using language like 'Register' or 'Sign up' instead of 'Create user'
+    = f.submit 'Register'
\ No newline at end of file
diff --git a/app/views/users/new_user.slim b/app/views/users/new_user.slim
deleted file mode 100644
index db0e2db..0000000
--- a/app/views/users/new_user.slim
+++ /dev/null
@@ -1,6 +0,0 @@
-h1
-  | New User
-
-= render 'form'
-
-= link_to 'Back', root_url
\ No newline at end of file
diff --git a/config/routes.rb b/config/routes.rb
index 1bf7079..004e8bc 100644
--- a/config/routes.rb
+++ b/config/routes.rb
@@ -2,8 +2,9 @@ Rails.application.routes.draw do
   resources :sessions, only: [:new, :create, :destroy]
   get '/sign-in' => 'sessions#new'
   delete '/sign-out' => 'sessions#destroy'
-  get '/new_user/create' => 'users#create'
-  get '/new_user' => 'users#new'
+
+  resources :users, only: [:new, :create]
+
   get 'static/home'

   root 'static#home'
```

**Comment by PStanos** [comment]
@chrisvfritz I've fixed things per your recommendations. I've also fixed similar Slim related issues, like the shortened div tags and removal of <br> tags (I used block elements to push things to new lines instead). Thanks for the great feedback and some shortcuts on things. I think what trips me up with Rails is that there is a lot of things that work "out of the box" like the repeat_password thing, which is much different than languages like C++ or C# where things are more explicitly defined. Anyway, great feedback!

**Comment by chrisvfritz** [comment]
:+1: Looks fantastic. And I understand where you're coming from regarding the out-ot-the-box magic. Rails' philosophy is *convention over configuration*, so the community decides what the most common configuration for 95% of apps might be and then makes that the default.

When I started using Rails, I was very, very uncomfortable with all the apparent magic happening behind the scenes. And I think it's a natural tendency for devs to want to build everything from scratch, so that they understand every part of their system. Over time though, you become more comfortable and grateful for these conventions and are better able to intuit them. You start to trust that logical things will happen most of the time and become more aware of the few "gotchas" that can crop up.

And when you're able to take what's in your head and make it real more quickly and easily than with any other framework (that I've used at least), it's hard to get too cranky. :smiley:
